src/hb-ot-metrics.cc

```diff
@@ -24,70 +24,130 @@
 
 #include "hb.hh"
 
-#ifndef HB_NO_METRICS
-
 #include "hb-ot-metrics.hh"
+#include "hb-ot-gasp-table.hh" // Just so we compile it; unused otherwise.
+#include "hb-ot-os2-table.hh"
+#include "hb-ot-post-table.hh"
+#include "hb-ot-hhea-table.hh"
 #include "hb-ot-var-mvar-table.hh"
 #include "hb-ot-face.hh"
 
+inline float
+_fix_ascender_descender (float value, hb_ot_metrics_t metrics_tag)
+{
+  if (metrics_tag == HB_OT_METRICS_HORIZONTAL_ASCENDER ||
+      metrics_tag == HB_OT_METRICS_VERTICAL_ASCENDER)
+    return fabs ((double) value);
+  if (metrics_tag == HB_OT_METRICS_HORIZONTAL_DESCENDER ||
+      metrics_tag == HB_OT_METRICS_VERTICAL_DESCENDER)
+    return -fabs ((double) value);
+  return value;
+}
+
+bool
+hb_ot_metrics_get_position_common (hb_font_t       *font,
+				   hb_ot_metrics_t  metrics_tag,
+				   hb_position_t   *position     /* OUT.  May be NULL. */)
+{
+  hb_face_t *face = font->face;
+#ifndef HB_NO_VAR
+#define GET_VAR hb_ot_metrics_get_variation (face, metrics_tag)
+#else
+#define GET_VAR 0
+#endif
+#define GET_METRIC(TABLE, ATTR, DIR) \
+  (face->table.TABLE->has_data () && \
+    (position && (*position = font->em_scalef_##DIR ( \
+      _fix_ascender_descender (face->table.TABLE->ATTR + GET_VAR, metrics_tag))), true))
+  if (metrics_tag == HB_OT_METRICS_HORIZONTAL_ASCENDER)
+    return (face->table.OS2->use_typo_metrics () && GET_METRIC (OS2, sTypoAscender, y)) ||
+	   GET_METRIC (hhea, ascender, y);
+  if (metrics_tag == HB_OT_METRICS_HORIZONTAL_DESCENDER)
+    return (face->table.OS2->use_typo_metrics () && GET_METRIC (OS2, sTypoDescender, y)) ||
+	   GET_METRIC (hhea, descender, y);
+  if (metrics_tag == HB_OT_METRICS_HORIZONTAL_LINE_GAP)
+    return (face->table.OS2->use_typo_metrics () && GET_METRIC (OS2, sTypoLineGap, y)) ||
+	   GET_METRIC (hhea, lineGap, y);
+  if (metrics_tag == HB_OT_METRICS_VERTICAL_ASCENDER)  return GET_METRIC (vhea, ascender, x);
+  if (metrics_tag == HB_OT_METRICS_VERTICAL_DESCENDER) return GET_METRIC (vhea, descender, x);
+  if (metrics_tag == HB_OT_METRICS_VERTICAL_LINE_GAP)  return GET_METRIC (vhea, lineGap, x);
```

Comment:
What the .... is this block of code pasta?!

Reply:
I can't use switch as it won't cover all the labels so a compiler warning happens, probably fixes if casted however.

Reply:
It is just separated from `hb_ot_metrics_get_position` for size consideration.